File: Python/Easy/242_valid_anagram.py
"""
242. Valid Anagram

Return whether t is an anagram of s — same characters, possibly reordered.

Problem Link: https://leetcode.com/problems/valid-anagram/

Examples:
    Example 1:
        Input: s = "racecar", t = "carrace"
        Output: true

    Example 2:
        Input: s = "jar", t = "jam"
        Output: false

Constraints:
    1 <= s.length, t.length <= 5 * 10^4
    s and t consist of lowercase English letters.
"""

from collections import Counter


class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        """
        Sort both strings and compare. Cheapest to write; pays a log factor
        you don't strictly need.

        Time complexity:  O(n log n)
        Space complexity: O(n) — sorted() materializes lists.
        """
        if len(s) != len(t):
            return False

        return sorted(s) == sorted(t)

    def isAnagramCounter(self, s: str, t: str) -> bool:
        """
        Character-frequency equality via Counter. Two counters are equal iff
        every key maps to the same count, so this captures anagram-ness in
        a single comparison.

        Time complexity:  O(n)
        Space complexity: O(k), k = alphabet size (26 here, so O(1)).
        """
        return Counter(s) == Counter(t)


if __name__ == "__main__":
    solution = Solution()

    test_cases = [
        ("anagram", "nagaram", True),
        ("rat", "car", False),
        ("listen", "silent", True),
        ("a", "a", True),
        ("ab", "ba", True),
        ("abc", "cba", True),
        ("abc", "abcd", False),
        ("racecar", "carrace", True),
        ("jar", "jam", False),
    ]

    for i, (s, t, expected) in enumerate(test_cases, 1):
        assert solution.isAnagram(s, t) == expected, f"isAnagram test {i} failed"
        assert solution.isAnagramCounter(s, t) == expected, f"isAnagramCounter test {i} failed"

    print("All test cases passed!")

File: Golang/Easy/242_valid_anagram.go
package main

import (
	"fmt"
	"slices"
)

// isAnagramSort sorts the rune slices of both strings and compares them.
// Cheapest to write; pays a log factor you don't strictly need.
//
// Time complexity:  O(n log n)
// Space complexity: O(n) — rune slices.
func isAnagramSort(s string, t string) bool {
	if len(s) != len(t) {
		return false
	}
	sRunes := []rune(s)
	tRunes := []rune(t)

	slices.Sort(sRunes)
	slices.Sort(tRunes)

	return slices.Equal(sRunes, tRunes)
}

// isAnagram uses a single rune-frequency map. Increment on s, decrement on
// t; bail the moment a count goes negative — that proves t has a character
// s doesn't (or has too many of one). The length pre-check rules out the
// "all positive at the end" case.
//
// Time complexity:  O(n)
// Space complexity: O(k), k = alphabet size.
func isAnagram(s string, t string) bool {
	if len(s) != len(t) {
		return false
	}

	counts := make(map[rune]int)

	for _, char := range s {
		counts[char]++
	}

	for _, char := range t {
		counts[char]--
		if counts[char] < 0 {
			return false
		}
	}
	return true
}

func main() {
	testCases := []struct {
		s        string
		t        string
		expected bool
	}{
		{"anagram", "nagaram", true},
		{"rat", "car", false},
		{"listen", "silent", true},
		{"a", "a", true},
		{"ab", "ba", true},
		{"abc", "cba", true},
		{"abc", "abcd", false},
		{"racecar", "carrace", true},
		{"jar", "jam", false},
	}

	funcs := []struct {
		label string
		fn    func(string, string) bool
	}{
		{"isAnagram    ", isAnagram},
		{"isAnagramSort", isAnagramSort},
	}

	for _, f := range funcs {
		for i, tc := range testCases {
			result := f.fn(tc.s, tc.t)
			status := "PASS"
			if result != tc.expected {
				status = "FAIL"
			}
			fmt.Printf("%s | Example %d | s=%q t=%q | Expected: %t | Result: %t | [%s]\n",
				f.label, i+1, tc.s, tc.t, tc.expected, result, status)
		}
	}
}
